### PR TITLE
Shuttle Supercruise + Orbital Mechanics

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## TODO

 - [x] Custom shuttles need to use this system. https://github.com/BeeStation/BeeStation-Hornet/pull/1398
 - [x] Finish forced autopilot ships
 - [x] Snowdin elevator can launch to space lol.
 - [x] Finish the ability to dock anywhere at designated points.
 - [x] Remove bluespace gigabeacon.
 - [x] Remove old shuttle docker.
 - [x] Fix autopilot not correcting tan() when to the bottom right of a target.
 - [ ] <s>Auto follow a shuttle on launch.</s>
 - [x] Add in a recall console that allows you to have the autopilot fly a shuttle to a specified dock. (Used for the syndicate base shuttle console)

## About The Pull Request

Revamps the old shuttle console and replaces it with a new shuttle console fit with manual shuttle piloting and gravitational bodies.

The primary object at the centre of the system is Lavaland. All objects orbit around lavaland (Sub orbits are possible, however).
At the start of the game, the position and velocities of different objects are calculated and they are set into orbit around lavaland.
The only forced orbit is the station, IE all other orbits are maintained through the gravity alone and no cheats.

Putting objects into a stable, circular orbit is pretty easy. Simply you need to set the velocity direction perpendicular to the direction towards the body of mass you are orbiting and the magnitude to sqrt(G(M+m)/r).

Current Exceptions to this system:
Any shuttle not controlled directly by players will not appear on the orbit map.
 - Emergency shuttle
 - Arrivals shuttle
 - Cargo Shuttle

Escape pods will use this system when launched manually to lavaland, however will perform the usual bluespace warp when evaccing normally.

Escape pods and the Syndicate Droppod cannot be manually controlled. These shuttles have autopilot forced on, meaning that you can only choose a destination and have the shuttle autopilot fly it with manual control completely disabled.

Direct Shuttle Changes:
Generally speaking, almost all shuttles have been directly buffed, since all shuttles (apart from ones that are autopilot forced) can go to any location on the map which allows for random, or anywhere docking.
The only exception is that the nuclear operatives declaration of war will prevent the use of the ability to dock anywhere.

Autopilot:
All ships are equipped with an advanced autopilot. When active the autopilot will automatically adjust the angle of the ship's thrust so that the velocity vector of the ship always points towards the target.

Location Types:
Beacon:
Beacons are random z-levels (space levels). Upon docking with a beacon you will be able to choose a location for the shuttle to land by using the standard shuttle camera docker.

Lavaland:
Lavaland is a planetary body and has the largest mass so that all objects orbit it. Upon colliding with lavaland, you will automatically dock with it. When docking with lavaland you can either travel to a random location, or one of the docks on lavaland if you have access to it. Random docking however is dangerous and can cause collisions with structures if you are unlucky.

Phobos:
Phobos is a moon in orbit around lavaland and is the home of the Centcom Z-Level. Docking here is impossible unless you have a valid port on the centcom Z-Level (Syndicate infiltrator).

Stars were also originally included but removed due to docking with SS13 while it was orbiting an object orbiting another object was near impossible. If SS13 collides with the sun, the station is destroyed and the round ends, however the sun doesn't exist anymore so this is not something to worry about.

Space Station 13:
The main station. Shuttles can either dock with a port or at any location on the SS13 Z-Level.

image

image

image

## Changelog
:cl:
add: Added orbits
add: Added shuttle supercruise flight
add: Shuttle autopilot
add: Orbital body gravitation
add: Added auto shuttles
add: Random landing on lavaland.
add: Any shuttle can choose a landing zone on the station or at random z-levels.
add: You can fly to random z-levels with shuttles on the orbital map.
refactor: Custom shuttles now use a subtype of the shuttle console
refactor: Custom shuttles use fuel at a constant rate
refactor: If a custom shuttle runs out of fuel it will be dumped on a random empty z-level.
del: Removed shuttle docker console
del: Removed shuttle navigation beacon
/:cl:

### CODE REVIEW

**Comment by ToasterBan** [comment]
b-based?

**Comment by TerryTehCoder** [comment]
This is actually awesome. Would love to see some like functionality added to like the Escape Pod Control consoles if emagged. Just override the autopilot and go on a freeflight in a Two Tile escape pod. 

We're really about to pull a Sevastopol out here and drive straight into lavaland.

**Comment by eeSPee** [comment]
I actually have something to compliment this in the works; Insta-shuttles.

Think mining bunkers, but they're one room of a shuttle, or a whole smaller shuttle.

**Comment by ZombiesVsPlants** [comment]
image

**Comment by ZombiesVsPlants** [comment]
The real question is, using custom shuttles, will 2 ships be able to rendezvous by matching velocity in orbit?

**Comment by ZombiesVsPlants** [comment]
I am also looking forward to actually using my custom shuttles to do space exploration without the tedium of placing bluespace beacons on every zlevel

**Comment by aidan-bowman** [comment]
ohmygod ohmygod ohmygod

spess dogfighting

if this gets merged i will put a bounty of 1 quintillion buckaroos for turrets on shuttles that a second pilot (or rather gunman) can operate for the absolutely sick star wars battle reenactment

honestly at this point you could probably use custom shuttles to replace lavaland mining

**Comment by mcmeiler** [comment]
> 
> 
> ohmygod ohmygod ohmygod
> 
> spess dogfighting
> 
> if this gets merged i will put a bounty of 1 quintillion buckaroos for turrets on shuttles that a second pilot (or rather gunman) can operate for the absolutely sick star wars battle reenactment
> 
> honestly at this point you could probably use custom shuttles to replace lavaland mining

Literally NSV

**Comment by aidan-bowman** [comment]
> > ohmygod ohmygod ohmygod
> > spess dogfighting
> > if this gets merged i will put a bounty of 1 quintillion buckaroos for turrets on shuttles that a second pilot (or rather gunman) can operate for the absolutely sick star wars battle reenactment
> > honestly at this point you could probably use custom shuttles to replace lavaland mining
> 
> Literally NSV

lmao really, never actually looked up what nsv is all i know is it's a third server on beestation

**Comment by mcmeiler** [comment]
> lmao really, never actually looked up what nsv is all i know is it's a third server on beestation

I hate advertising and making off-topic comments, but you should try nsv if it goes up lol

**Comment by eeSPee** [comment]
But in all seriousness, this kinda raises more serious concerns:
* What if cult does cult shuttle and cruises through space?
* What if you're antag and your dude is off in space somewhere
* Is it powergaming on MRP to have a fully armed shuttle?
* Will this not drive the gameplay away from the SS and further into space?

**Comment by PowerfulBacon** [comment]
>     * What if cult does cult shuttle and cruises through space?

Cult cannot use runes in space or on the transit z-level.

>     * What if you're antag and your dude is off in space somewhere

Then, like currently, you will not be able to assassinate them unless you go to their position. If you know what z-level they are on it is much easier to reach them now since you can hijack an unused shuttle like the security shuttle and fly it out.

>     * Is it powergaming on MRP to have a fully armed shuttle?

Shuttles don't have weapons.

>     * Will this not drive the gameplay away from the SS and further into space?

The shuttles can go to lavaland and space, which are already off-station locations accessible by shuttles. Could encourage people to leave station more since its easier with custom shuttles, but the better piloting is worth it.

File: code/__DEFINES/orbit_defines.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+#define GRAVITATIONAL_CONSTANT 1
+
+//Once the acceleration towards this object is smaller than this value, it will be ignored.
+#define MINIMUM_EFFECTIVE_GRAVITATIONAL_ACCEELRATION 0.0001
+
+#define ORBITAL_UPDATE_RATE 1 SECONDS	//10 stupidseconds
```
Comment (ike709) [inline code/__DEFINES/orbit_defines.dm]:
```suggestion
#define ORBITAL_UPDATE_RATE (1 SECONDS)	//10 stupidseconds
```
Just to be safe

File: code/modules/shuttle/computer.dm [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/obj/machinery/computer/shuttle
+/*/obj/machinery/computer/shuttle_flight
```
Comment (ike709) [inline code/modules/shuttle/computer.dm]: Commented-out code

File: code/modules/shuttle/custom_shuttle.dm [hunk 3]
```diff
@@ -1,124 +1,132 @@
...
+	calculated_acceleration = (calculated_dforce*1000) / (calculated_mass*100)
+
+/obj/machinery/computer/shuttle_flight/custom_shuttle/proc/check_stranded()
+	if(!calculated_engine_count && shuttleObject)
+		say("Fuel reserves depleted, dropping out of supercruise.")
+		if(!shuttleObject.docking_target)
+			if(shuttleObject.can_dock_with)
+				shuttleObject.docking_target = shuttleObject.can_dock_with
+			else
+				//Send shuttle object to random location
```
Comment (ike709) [inline code/modules/shuttle/custom_shuttle.dm:97]: A bold choice...

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+/datum/orbital_object/star/collision(datum/orbital_object/other)
+	//You got lucky this time
+	if(other.collision_ignored || collision_ignored)
+		return
+	//You didnt get lucky this time
+	message_admins("Orbital body [other.name] collided with the star [name] and was destroyed.")
+	log_game("Orbital body [other.name] collided with the star [name] and was destroyed.")
+	//Rip.
+	if(istype(other, /datum/orbital_object/star))
+		//Make a black hole or something
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm:17]: OwO

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/space_station.dm [hunk 5]
```diff
@@ -0,0 +1,16 @@
+/datum/orbital_object/z_linked/station
+	name = "Space Station 13"
+	mass = 0
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/space_station.dm:3]: No mass?

Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/space_station.dm:3]: No, mass only determines gravitational attraction of other objects towards it

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm:17]: Stars were removed due to the fact that attempting to orbit the station while its orbiting lavaland orbiting a star was a near impossible feat

File: code/modules/shuttle/custom_shuttle.dm (hunk 3 above)
Comment (PowerfulBacon) [reply code/modules/shuttle/custom_shuttle.dm:97]: It only sends them to beacons, which are random empty / space ruin z-levels, unless the shuttle is above a docking location in which case its sent to that z-level.

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm [hunk 6]
```diff
@@ -0,0 +1,17 @@
+
+/datum/orbital_map
+	//the primary star. Set to be lavaland by default.
+	var/datum/orbital_object/center = null
+	//A list of all bodies
+	var/list/datum/orbital_object/bodies = list()
+
+//Returns a list of gravitationally relevant bodies.
+/datum/orbital_map/proc/get_relevnant_bodies(datum/orbital_object/source)
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm:9]: > relevnant

Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm:9]: Once the distance is high enough, the acceleration towards the gravitational body will be less than 0.0001, at which case the gravity calculations are ignored.

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/phobos.dm [hunk 7]
```diff
@@ -0,0 +1,11 @@
+//Centcom Z-Level.
+//Syndicate infiltrator level.
+//TODO: Normal ships arriving here will crash on a snowy z-level.
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/phobos.dm]: neat

Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/phobos.dm]: uhhhhh idk if its worth making an extra z-level that wont be used outside of this

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm (hunk 6 above)
Comment (ike709) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm:9]: I was referring to the typo

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm (hunk 4 above)
Comment (ike709) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm:17]: sad

**Comment by ike709** [comment]
Gave everything except the math a quick once-over. No obvious issues.

Also, do virtual zlevels need to be tested and merged before this?

**Comment by PowerfulBacon** [comment]
No virtual z-levels are independant

**Comment by PowerfulBacon** [comment]
Well the autopilot still has some issues converting vectors to angles it seems

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm (hunk 4 above)
Comment (utilbird) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm:17]: how can you orbit a station

Comment (aidan-bowman) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_objects/star.dm:17]: i mean, you can orbit ANYTHING. it does raise the question of how close orbiting a man-made station with a man-made spaceship would be to a binary system

File: code/modules/shuttle/super_cruise/shuttle_components/shuttle_console.dm [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+
+/obj/machinery/computer/shuttle_flight/Initialize(mapload, obj/item/circuitboard/C)
+	. = ..()
+	valid_docks = params2list(possible_destinations)
+	shuttleId = shuttleId
+	shuttlePortId = "[shuttleId]_custom"
+
+/obj/machinery/computer/shuttle_flight/Destroy()
+	. = ..()
+	shuttleObject = null
```
Comment (utilbird) [inline code/modules/shuttle/super_cruise/shuttle_components/shuttle_console.dm:37]: Should this be `QDEL_NULL`?

Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/shuttle_components/shuttle_console.dm:37]: No, other consoles reference this object

**Comment by PowerfulBacon** [comment]
> Well the autopilot still has some issues converting vectors to angles it seems

Y direction is flipped on the UI which calculations do not consider

**Comment by eeSPee** [comment]
Rework Gateway missions to be shuttle accessible?

**Comment by eeSPee** [comment]
> Then, like currently, you will not be able to assassinate them unless you go to their position. If you know what z-level they are on it is much easier to reach them now since you can hijack an unused shuttle like the security shuttle and fly it out.

I think many problems could arise from shuttles not having a built-in GPS system and becoming hard to track. This may be just my opinion, but I think shuttle consoles should have an inbuilt GPS.

> The shuttles can go to lavaland and space, which are already off-station locations accessible by shuttles. Could encourage people to leave station more since its easier with custom shuttles, but the better piloting is worth it.

Yes, for now, but this could also be the reason this doesn't add much and will be a widely ignored feature.

**Comment by PowerfulBacon** [comment]
You cant ignore this if you use a shuttle (mining, sec prison, nukies etc.)

**Comment by PowerfulBacon** [comment]
image

```
proc name: random drop (/obj/machinery/computer/shuttle_flight/proc/random_drop
proc name: Normalize (/datum/orbital_vector/proc/Normalize)
```

Result from autopilot target being selected as self.

**Comment by PowerfulBacon** [comment]
Stealth ships cause all ship UIs to break

**Comment by Boxlta** [comment]
Would this tie into mechanics with #2551? It seems like those two PRs could go really well together.

**Comment by PowerfulBacon** [comment]
> Would this tie into mechanics with #2551? It seems like those two PRs could go really well together.

I would like to salvage parts of #2551 specifically the subsystem (This PR is flight and shuttles only)

File: code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm (hunk 6 above)
Comment (PowerfulBacon) [reply code/modules/shuttle/super_cruise/orbital_map_components/orbital_map.dm:9]: damn it

**Comment by PowerfulBacon** [comment]
The window for the thing is too big

**Comment by eeSPee** [comment]
Does this PR also contain the randomly generated dungeons?

**Comment by PowerfulBacon** [comment]
no

**Comment by ike709** [comment]
So should this still be a thing that gets merged before Part 2 or are you just rolling all of that into one PR?

**Comment by PowerfulBacon** [comment]
> 
> 
> So should this still be a thing that gets merged before Part 2 or are you just rolling all of that into one PR?

rolling it into 1

**Pull request closed without merging**